dali/kernels/shape.h

```diff
@@ -0,0 +1,419 @@
+    TensorShape<DynamicDimensions> result;
+    result.shape.resize(count);
+    for (int i = 0; i < count; i++) {
+      result[i] = (*this)[i];
+    }
+    return result;
+  }
+
+  TensorShape<DynamicDimensions> last(int count) {
+    assert(0 <= count && count <= size() &&
+           "Number of elements in subshape must be between 0 and size()");
+    TensorShape<DynamicDimensions> result;
+    result.shape.resize(count);
+    int start_offset = size() - count;
+    for (int i = 0; i < count; i++) {
+      result[i] = (*this)[start_offset + i];
+    }
+    return result;
+  }
+};
+
+template <int ndim>
+struct TensorShape : public TensorShapeBase<std::array<int64_t, ndim>> {
+  using Base = TensorShapeBase<std::array<int64_t, ndim>>;
+  TensorShape(const std::array<int64_t, ndim> &s) : Base(s) {}
+  // We just want zero-initialization
+  TensorShape() = default;
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : Base(typename Base::container_type{i0, int64_t{s}...}) {
+    static_assert(sizeof...(Ts) == ndim - 1, "Number of shapes passed must match ndim");
+  }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> first() {
+    static_assert(0 <= other_ndim && other_ndim <= ndim,
+                  "Number of elements in subshape must be between 0 and size()");
+    TensorShape<other_ndim> result;
+    for (int i = 0; i < other_ndim; i++) {
+      result[i] = (*this)[i];
+    }
+    return result;
+  }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> last() {
+    static_assert(0 <= other_ndim && other_ndim <= ndim,
+                  "Number of elements in subshape must be between 0 and size()");
+    TensorShape<other_ndim> result;
+    int start_offset = ndim - other_ndim;
+    for (int i = 0; i < other_ndim; i++) {
+      result[i] = (*this)[start_offset + i];
+    }
+    return result;
+  }
+};
+
+template <int left_ndim, int right_ndim>
+bool operator==(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  if (left.size() != right.size()) {
+    return false;
+  }
+  int size = left.size();
+  for (int i = 0; i < size; i++) {
+    if (left[i] != right[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator!=(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  return !(left == right);
+}
+
+template <int left_ndim, int right_ndim>
+typename std::enable_if<left_ndim == DynamicDimensions || right_ndim == DynamicDimensions,
+                        TensorShape<DynamicDimensions>>::type
+shape_cat(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  TensorShape<DynamicDimensions> result;
+  int total_size = left.size() + right.size();
+  result.shape.resize(total_size);
```

Comment:
If we add `resize` to `TensorShape`, we won't need to have two distinct functions.

Reply:
Done